src/gc-pages.c

```diff
@@ -96,113 +84,184 @@ static void jl_gc_alloc_region(region_t *region)
+                            assert(info.meta->data);
+                            // new pages available starting at min of lb and region_i32
+                            if (memory_map.lb < info.region_i32)
+                                memory_map.lb = info.region_i32;
+                            if (info.region1->lb < info.region1_i32)
+                                info.region1->lb = info.region1_i32;
+                            if (info.region0->lb < info.region0_i32)
+                                info.region0->lb = info.region0_i32;
+                            goto have_free_page;
+                        }
+                    }
+                    info.region1->freemap0[info.region1_i32] &= ~(uint32_t)(1 << info.region1_i); // record that this was full
+                }
+            }
+            memory_map.freemap1[info.region_i32] &= ~(uint32_t)(1 << info.region_i); // record that this was full
         }
-        break;
     }
-    if (__unlikely(region_i >= REGION_COUNT)) {
-        JL_UNLOCK_NOGC(&pagealloc_lock);
-        jl_throw(jl_memory_exception);
+
+    // no existing pages found, allocate a new one
+    {
+        jl_gc_pagemeta_t *meta = jl_gc_alloc_new_page();
+        info = page_metadata_ext(meta->data);
+        assert(meta == info.meta);
     }
-    if (region->lb < i)
-        region->lb = i;
-    if (region->ub < i)
-        region->ub = i;
-
-#if defined(_COMPILER_MINGW_)
-    int j = __builtin_ffs(~region->allocmap[i]) - 1;
-#elif defined(_COMPILER_MICROSOFT_)
-    unsigned long j;
-    _BitScanForward(&j, ~region->allocmap[i]);
-#else
-    int j = ffs(~region->allocmap[i]) - 1;
-#endif
+    // new pages available starting at max of lb and region_i32
+    if (memory_map.lb > info.region_i32)
+        memory_map.lb = info.region_i32;
+    if (info.region1->lb > info.region1_i32)
+        info.region1->lb = info.region1_i32;
+    if (info.region0->lb > info.region0_i32)
+        info.region0->lb = info.region0_i32;
+
+have_free_page:
+    if (memory_map.ub < info.region_i32)
+        memory_map.ub = info.region_i32;
+    if (info.region1->ub < info.region1_i32)
+        info.region1->ub = info.region1_i32;
+    if (info.region0->ub < info.region0_i32)
+        info.region0->ub = info.region0_i32;
+
+    // mark this entry as in-use and not free
+    info.region0->freemap[info.region0_i32] &= ~(uint32_t)(1 << info.region0_i);
+    info.region0->allocmap[info.region0_i32] |= (uint32_t)(1 << info.region0_i);
+    info.region1->allocmap0[info.region1_i32] |= (uint32_t)(1 << info.region1_i);
+    memory_map.allocmap1[info.region_i32] |= (uint32_t)(1 << info.region_i);
 
-    region->allocmap[i] |= (uint32_t)(1 << j);
-    void *ptr = region->pages[i * 32 + j].data;
 #ifdef _OS_WINDOWS_
-    VirtualAlloc(ptr, GC_PAGE_SZ, MEM_COMMIT, PAGE_READWRITE);
+    VirtualAlloc(meta->data, GC_PAGE_SZ, MEM_COMMIT, PAGE_READWRITE);
```

Comment:
It looks like this is using `meta` outside of its scope (defined on line 178 inside a block) which is causing the Windows failures.